cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/OutboundHandler.java

```diff
@@ -39,9 +42,15 @@ public OutboundHandler(Channel inboundChannel) {
   }
 
   @Override
-  public void messageReceived(ChannelHandlerContext ctx, final MessageEvent e) throws Exception {
+  public void messageReceived(final ChannelHandlerContext ctx, final MessageEvent e) throws Exception {
     ChannelBuffer msg = (ChannelBuffer) e.getMessage();
-    inboundChannel.write(msg);
+
+    Channels.write(inboundChannel, msg).addListener(new ChannelFutureListener() {
+      @Override
+      public void operationComplete(ChannelFuture future) throws Exception {
+        OutboundHandler.super.messageReceived(ctx, e);
```

Comment:
its calling the super's implementation of the method (which sends the message further upstream).
We do it in a callback, to make it happen after the write happens to the inbound channel.